[Dashboard]: add send telemetry to admin settings

MrSimonEmms opened the pull request:
## Description
<!-- Describe your changes in detail -->

This adds the "send telemetry" checkbox to the Admin dashboard and sets the default option to `true` - the front-end for #7591

It also moves the `guardAdminAccess` method from the EE `gitpod-server-impl.ts` to the one in `src`. This is because we're going to open up more dashboard administrative options to users regardless of whether they have an enterprise license or not - clearly, the option to disable telemetry should be available to all users.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes #7573

## How to test
<!-- Provide steps to test this PR -->
Go to the [Dashboard](https://sje-installation-admin-dashboard.staging.gitpod-dev.com/workspaces) -> Settings and click the "send telemetry" checkbox

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
[Dashboard]: add send telemetry to admin settings
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

roboquat commented:
@MrSimonEmms: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** and additionally assign **jldec** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @jldec` in a comment when ready.

Associated issue: *#7573*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/installation-telemetry/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/installation-telemetry/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**
- **[installer/pkg/components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/pkg/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jldec","mrsimonemms"]} -->

MrSimonEmms commented:
/werft run

MrSimonEmms commented:
/werft run

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** and additionally assign **jldec** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @jldec` in a comment when ready.

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/installation-telemetry/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/installation-telemetry/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**
- **[installer/pkg/components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/pkg/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jldec","mrsimonemms"]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jldec** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @jldec` in a comment when ready.

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jldec"]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jldec** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @jldec` in a comment when ready.

Associated issue: *#7573*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jldec"]} -->

MrSimonEmms commented:
/assign @geropl

MrSimonEmms commented:
/assign @gitpod-io/engineering-webapp

geropl commented on components/dashboard/src/admin/Settings.tsx:
```diff
@@ -0,0 +1,33 @@
+        <div>
+            <PageWithSubMenu subMenu={adminMenu} title="Settings" subtitle="Configure settings for your Gitpod cluster.">
+                <CheckBox
+                    title="Send anonymous service ping to Gitpod"
```
1st questions that pops up here is: "What exactly is sent?" Either explained here, or with a ref to docs.

MrSimonEmms replied:
Good question. We had a @gitpod-io/engineering-self-hosted [discussion on this](https://gitpod.slack.com/archives/C01KLC56NP7/p1642696983023300) and agreed that wording was enough to be going on with.

Is there a way of putting a link to the [privacy policy](https://www.gitpod.io/privacy) in the description?

geropl commented on components/gitpod-protocol/src/admin-protocol.ts:
```diff
@@ -29,6 +29,9 @@ export interface AdminServer {
     adminAddStudentEmailDomain(userId: string, domain: string): Promise<void>;
     adminGrantExtraHours(userId: string, extraHours: number): Promise<void>;
+
+    adminSettingsGetTelemetry(): Promise<boolean>
```
Instead of having a method here for every boolean flag, we should implement something akin [`updateLoggedInUser`](https://github.com/gitpod-io/gitpod/blob/0951ced5ebfb2e1a822b7c3eaa4855f7cad7724c/components/server/src/workspace/gitpod-server-impl.ts#L283-L298).

This avoids the necessity to update the API surface and implementation on every change.

geropl replied:
Note that it's important to have an explicit allowlist for fields to update, otherwise this is an attack vector.

One way to solve this without further duplication would be to change `installation-admin-protocol.ts` to sth like this:
```
/**
 * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
 * Licensed under the GNU Affero General Public License (AGPL).
 * See License-AGPL.txt in the project root for license information.
 */


const InstallationAdminSettingsPrototype = {
    sendTelemetry: true,
};
export type InstallationAdminSettings = typeof InstallationAdminSettingsPrototype;
export namespace InstallationAdminSettings {
    export function fields(): (keyof InstallationAdminSettings)[] {
        return Object.keys(InstallationAdminSettingsPrototype) as (keyof InstallationAdminSettings)[];
    }
}

export interface InstallationAdmin {
    id: string;
    settings: InstallationAdminSettings;
}
export namespace InstallationAdmin {
    export function createDefault(): InstallationAdmin {
        return {
            id: "",
            settings: {
                ...InstallationAdminSettingsPrototype,
            }
        };
    }
}
```
And use the `InstallationAdminSettings.fields()` method to iterate the shape's properties like in `updateLggedInUser` with the explicit list.

geropl commented on components/gitpod-protocol/src/gitpod-service.ts:
```diff
@@ -131,6 +131,10 @@ export interface GitpodServer extends JsonRpcServer<GitpodClient>, AdminServer,
     deleteTeam(teamId: string, userId: string): Promise<void>;
 
+    // Admin Settings
+    adminSettingsGetTelemetry(): Promise<boolean>;
```
See [comment above](https://github.com/gitpod-io/gitpod/pull/7732/files#r791663353)

geropl commented on components/gitpod-db/src/installation-admin-db.ts:
```diff
@@ -9,4 +9,5 @@ import { InstallationAdmin } from "@gitpod/gitpod-protocol";
 export const InstallationAdminDB = Symbol('InstallationAdminDB');
 export interface InstallationAdminDB {
     getTelemetryData(): Promise<InstallationAdmin>;
+    setSendTelemetry(value: boolean): Promise<void>;
```
See [comment below](https://github.com/gitpod-io/gitpod/pull/7732/files#r791663353).

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -564,6 +565,15 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         await this.internalStopWorkspaceInstance(ctx, instance.id, instance.region, policy);
     }
 
+    protected async guardAdminAccess(method: string, params: any, requiredPermission: PermissionName) {
```
:+1:

gtsiolis commented:
Looking at UX changes in this now! 👀 

Thanks for the ping, @geropl. 🏓

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["geropl"]} -->

gtsiolis commented on components/dashboard/src/admin/Settings.tsx:
```diff
@@ -0,0 +1,33 @@
+            <PageWithSubMenu subMenu={adminMenu} title="Settings" subtitle="Configure settings for your Gitpod cluster.">
+                <CheckBox
+                    title="Send anonymous service ping to Gitpod"
+                    desc="This is used to provide insights on how you use your cluster so we can provide a better overall experience."
```
**suggestion**: If we change the title as described above we could go with one of the following proposals depending if we have any relevant docs in place:

Option :a:
```suggestion
                    desc={<span>To help improve Gitpod, service ping will periodically collect anonymous usage information. <a className="gp-link" href="https://www.gitpod.io/docs/LINK-TO-DOCS-OR-PRIVACY-POLICY">Learn more</a></span>}
```

Option :b:
```suggestion
                    desc={<span>To help improve Gitpod, service ping will periodically collect anonymous usage information. Read our <a className="gp-link" href="https://www.gitpod.io/privacy">Read our Privavy Policy</a></span>}
```

gtsiolis commented on components/dashboard/src/App.tsx:
```diff
@@ -52,6 +53,7 @@ const InstallGitHubApp = React.lazy(() => import(/* webpackPrefetch: true */ './
 const FromReferrer = React.lazy(() => import(/* webpackPrefetch: true */ './FromReferrer'));
 const UserSearch = React.lazy(() => import(/* webpackPrefetch: true */ './admin/UserSearch'));
 const WorkspacesSearch = React.lazy(() => import(/* webpackPrefetch: true */ './admin/WorkspacesSearch'));
+const AdminSettings = React.lazy(() => import(/* webpackPrefetch: true */ './admin/Settings'));
```
**issue(non-blocking)**: Just wanted to add a comment that I could not access the admin panel evern after trying to add a user as an admin using this documentation[[1](https://www.notion.so/gitpod/How-we-develop-e8ce7e562478458a9223eb9b797415b2#d0f67538075f4050ba7c2551c7160f3b)].

MrSimonEmms replied:
I'm not sure about adding a user as admin. In my environment where my user is an admin by default, it shows ok. As you say, I think it's probably ok as there will always be the initial admin. We can always raise as a bugfix if a wider problem

gtsiolis replied:
Yes, the first user always becomes an admin. That was more like an observation that I could not test this in the preview environment properly either because the script to make someone admin does not work or because we could improve our code review process.

MrSimonEmms replied:
I've raised PR #7840 with a fix for leeway

gtsiolis reviewed (approved):
Thanks, @MrSimonEmms! 🔮 

Left some minor comments. ✔️ 

Approving to unblock merging but holding in case we'd like to address any of the UX feedback added below. 🏁 

/hold

roboquat commented:
LGTM label has been added.  <details>Git tree hash: 863e9a730d4ac464a78bb6ae6afee173f407b80c</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7732#pullrequestreview-862274333" title="Approved">gtsiolis</a>*

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7732#pullrequestreview-862274333" title="Approved">gtsiolis</a>*

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
New changes are detected. LGTM label has been removed.

MrSimonEmms commented:
@geropl I've updated as requested. I believe that's as requested, so please let me know if I've misunderstood any of your change requests.

@gtsiolis I've added in the link - thank you. I've slightly changed the text so it's more in keeping with what @gitpod-io/engineering-self-hosted agreed on.

![image](https://user-images.githubusercontent.com/275848/151019185-e2705a2f-4a3a-4ac8-9240-ab8570401720.png)

MrSimonEmms commented:
/unhold

MrSimonEmms commented:
/hold just realised that there's no way to the Admin section if you're not an admin

MrSimonEmms commented on components/dashboard/src/admin/Settings.tsx:
```diff
@@ -0,0 +1,33 @@
+        <div>
+            <PageWithSubMenu subMenu={adminMenu} title="Settings" subtitle="Configure settings for your Gitpod cluster.">
+                <CheckBox
+                    title="Send anonymous service ping to Gitpod"
```
Yes we can. At this stage, we only send a randomly generated ID and the version tag - see the [code](https://github.com/gitpod-io/gitpod/blob/main/components/installation-telemetry/cmd/send.go#L53)

gtsiolis replied:
Nice! Thanks for confirming, @MrSimonEmms!

MrSimonEmms commented on components/dashboard/src/admin/Settings.tsx:
```diff
@@ -0,0 +1,33 @@
+        <div>
+            <PageWithSubMenu subMenu={adminMenu} title="Settings" subtitle="Configure settings for your Gitpod cluster.">
+                <CheckBox
+                    title="Send anonymous service ping to Gitpod"
```
#7820

MrSimonEmms commented:
/unhold

gtsiolis commented:
@MrSimonEmms could you make me an admin in the preview env so that I can see the UX changes in action?

MrSimonEmms commented:
/werft run

MrSimonEmms commented:
@gtsiolis done. I've added you as well Gero

gtsiolis commented on components/dashboard/src/admin/Settings.tsx:
```diff
@@ -0,0 +1,36 @@
+    return (
+        <div>
+            <PageWithSubMenu subMenu={adminMenu} title="Settings" subtitle="Configure settings for your Gitpod cluster.">
+                <CheckBox
```
**suggestion**: What do you think of adding a section header for this setting as we do for all other settings across the product? Not sure if the code suggestion here would work as expected. 😇 

```suggestion
                <h3>Usage Statistics</h3>
                <CheckBox
```

gtsiolis commented on components/dashboard/src/admin/Settings.tsx:
```diff
@@ -0,0 +1,36 @@
+            <PageWithSubMenu subMenu={adminMenu} title="Settings" subtitle="Configure settings for your Gitpod cluster.">
+                <CheckBox
+                    title="Enable Service Ping"
+                    desc={<span>This is used to provide insights on how you use your cluster so we can provide a better overall experience. Read our <a className="gp-link" href="https://www.gitpod.io/privacy">Privacy Policy</a></span>}
```
**suggestion(non-blocking)**: To re-use the pattern of the rest of help links (_Learn more_ links) across the product we could hyperlink the whole action here. What do you think? 

```suggestion
                    desc={<span>This is used to provide insights on how you use your cluster so we can provide a better overall experience. <a className="gp-link" href="https://www.gitpod.io/privacy">Read our privacy policy</a></span>}
```

MrSimonEmms replied:
You're the boss when it comes to the front-end - I'm happy to oblige with this

gtsiolis reviewed (approved):
UX works like a charm, @MrSimonEmms! 

Thanks for making me admin to see this in actions and opening the follow up PR in https://github.com/gitpod-io/gitpod/pull/7840. 🏀 

Left two minor suggestions but feel free to ignore these if needed.

| BEFORE | AFTER |
|-|-|
| <img width="1440" alt="Screenshot 2022-01-26 at 2 28 54 PM (2)" src="https://user-images.githubusercontent.com/120486/151163570-7f17f9cf-b9b8-42eb-81f0-013007c15a87.png"> | <img width="1440" alt="Screenshot 2022-01-26 at 2 28 48 PM (2)" src="https://user-images.githubusercontent.com/120486/151163574-7431beb0-8203-4619-a048-c0733a2036b5.png"> |

Approving to unblock merging but holding in case we'd like to address any of the pending comments. Your call. 🏓 

/hold

roboquat commented:
LGTM label has been added.  <details>Git tree hash: e094102bbfceed842f01750fef0ae4d1ecfb1e0a</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7732#pullrequestreview-863474737" title="Approved">gtsiolis</a>*

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -2006,6 +2017,24 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
+    async adminUpdateSettings(ctx: TraceContext, settings: InstallationAdminSettings): Promise<void> {
+        traceAPIParams(ctx, {});
+
+        await this.guardAdminAccess("adminSettingsGetTelemetry", {}, Permission.ADMIN_API);
```
This should read `adminUpdateSettings` I guess. :+1:

MrSimonEmms replied:
Well spotted - will change

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -2006,6 +2017,24 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         throw new ResponseError(ErrorCodes.EE_FEATURE, `Admin support is implemented in Gitpod's Enterprise Edition`);
     }
 
+    async adminGetSettings(ctx: TraceContext): Promise<InstallationAdminSettings> {
```
:heart:

geropl commented on components/gitpod-db/src/typeorm/installation-admin-db-impl.ts:
```diff
@@ -55,4 +49,14 @@ export class TypeORMInstallationAdminImpl implements InstallationAdminDB {
+    async setSettings(settings: InstallationAdminSettings): Promise<void> {
+        const record = await this.getData();
+        record.settings = {
+            ...settings,
```
@MrSimonEmms This `...` might be an attack vector, depending on the implementation of `repo.save(...)`. (we had one with `.update` last fall).

I suggest we:
 - keep the spread operator (`...`) here
 - add an explicit loop over all fields in `api.adminUpdateSettings` (cmp. comment below)

geropl commented on components/gitpod-db/src/typeorm/installation-admin-db-impl.ts:
```diff
@@ -55,4 +49,14 @@ export class TypeORMInstallationAdminImpl implements InstallationAdminDB {
+        }
+
+        const repo = await this.getInstallationAdminRepo();
+        repo.save(record);
```
`await` missing.

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -2006,6 +2017,24 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
+
+        await this.guardAdminAccess("adminSettingsGetTelemetry", {}, Permission.ADMIN_API);
+
+        await this.installationAdminDb.setSettings(settings);
```
Instead of calling this directly, we need to check the properties we expect in `settings` explicitly:
```
        const newSettings: Partial<Settings> = {};
        for (const p of InstallationAdminSettings.fields()) {
            if (p in settings) {
                (newSettings[p] as any) = settings[p];
            }
        }
       await this.installationAdminDb.setSettings(newSettings);
```

roboquat commented:
New changes are detected. LGTM label has been removed.

geropl commented on components/server/src/installation-admin/installation-admin-controller.ts:
```diff
@@ -16,7 +16,7 @@ export class InstallationAdminController {
-            const data = await this.installationAdminDb.getTelemetryData();
+            const data = await this.installationAdminDb.getData();
 
             res.send(data);
```
nit: `res.status(200).json(data);` also sets the proper content header. Feel free to ignore, as you're managing the only client here :+1:

geropl commented on components/gitpod-db/src/installation-admin-db.ts:
```diff
@@ -4,9 +4,10 @@
 export interface InstallationAdminDB {
-    getTelemetryData(): Promise<InstallationAdmin>;
+    getData(): Promise<InstallationAdmin>;
+    setSettings(settings: InstallationAdminSettings): Promise<void>;
```
nit: this requires only `settings: Partial<InstallationAdmingSettings>`

geropl reviewed (approved):
Nice @MrSimonEmms :rocket: 

I left two more nits, but feel free to ignore to proceed here!

roboquat commented:
LGTM label has been added.  <details>Git tree hash: bbd9a19ca4bec142f1ed26d44a299b1131e68f82</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7732#pullrequestreview-863630053" title="Approved">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7732#pullrequestreview-863474737" title="Approved">gtsiolis</a>*

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
New changes are detected. LGTM label has been removed.

MrSimonEmms commented:
@geropl nits fixed

geropl commented:
/lgtm

roboquat commented:
LGTM label has been added.  <details>Git tree hash: 6271ab208f8968504a76720a3a148d9360d078f9</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7732#issuecomment-1022271166" title="LGTM">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7732#pullrequestreview-863474737" title="Approved">gtsiolis</a>*

Associated issue: *#7591*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

MrSimonEmms commented:
@geropl @gtsiolis thank you both for your help/patience on these PRs 🙏🏻  🎉

MrSimonEmms commented:
/unhold
